### ISSUE TITLE
About inputs[3] maxSeqlen in demo/Bert/builder_varseqlen.py

### ISSUE THREAD

**Issue by pogevip** [NVIDIA/TensorRT#1093]
About  inputs[3] maxSeqlen in builder_varseqlen.py

 inputs[0]  input_ids
 inputs[1]  segment_ids
 inputs[2]  cu_seqlens
 inputs[3]  max_seqlen

inputs[3] maxSeqlen in builder_varseqlen.py
https://github.com/NVIDIA/TensorRT/blob/master/demo/BERT/builder_varseqlen.py#L523

Only inputs[3] 's shape is used in the plugin.
embedding
https://github.com/NVIDIA/TensorRT/blob/master/plugin/embLayerNormPlugin/embLayerNormVarSeqlenPlugin.cpp#L266 
QKV
https://github.com/NVIDIA/TensorRT/blob/master/plugin/bertQKVToContextPlugin/qkvToContextPlugin.cpp#L889

I can not find any assignment to intput[3]
https://github.com/NVIDIA/TensorRT/blob/master/demo/BERT/inference_varseqlen.py#L136

What is the meaning of  input[3] max_seqlen？May anyone give me an example ?

**Comment by pogevip** [comment]
@ttyio Could you help me?

**Comment by ttyio** [comment]
Hello @pogevip , this is mainly used as hint to the multihead attention kernel about the problem size (max combined sequence length), so that plugin can select the kernel with best performance. 
This is a shape tensor, only the dimensions are used, and the content in the tensor is dummy values, that's the reason why we did not initialize it in the script.

**Comment by pogevip** [comment]
> Hello @pogevip , this is mainly used as hint to the multihead attention kernel about the problem size (max combined sequence length), so that plugin can select the kernel with best performance.
> This is a shape tensor, only the dimensions are used, and the content in the tensor is dummy values, that's the reason why we did not initialize it in the script.

@ttyio 
So it could be anything, right? 
For example  [1,1,1,···,1] ,I just make sure that its shape[0] is equal to the max sequence length I want to specify?

**Comment by ttyio** [comment]
> > Hello @pogevip , this is mainly used as hint to the multihead attention kernel about the problem size (max combined sequence length), so that plugin can select the kernel with best performance.
> > This is a shape tensor, only the dimensions are used, and the content in the tensor is dummy values, that's the reason why we did not initialize it in the script.
> 
> @ttyio
> So it could be anything, right?
> For example [1,1,1,···,1] ,I just make sure that its shape[0] is equal to the max sequence length I want to specify?

Yes @pogevip

**Comment by pogevip** [comment]
> > > Hello @pogevip , this is mainly used as hint to the multihead attention kernel about the problem size (max combined sequence length), so that plugin can select the kernel with best performance.
> > > This is a shape tensor, only the dimensions are used, and the content in the tensor is dummy values, that's the reason why we did not initialize it in the script.
> > 
> > 
> > @ttyio
> > So it could be anything, right?
> > For example [1,1,1,···,1] ,I just make sure that its shape[0] is equal to the max sequence length I want to specify?
> 
> Yes @pogevip

Thank you @ttyio

**Comment by pogevip** [comment]
@ttyio Excuse me, I have one more problem.

I have got an engine  by running builder_varseqlen.py.
Now I want to deploy to the server by TritonServer, but I get this error in a client:
```
InferenceServerException: input 'max_seqlen' batch size does not match other inputs for 'varbert
```
The error could come from this code.
https://github.com/triton-inference-server/server/blob/master/src/core/infer_request.cc#L581

The following is mentioned in the code：
```
    // Model does support Triton-style batching so each input tensor
    // must have the same first dimension which is the batch
    // size. Adjust the shape of the input tensors to remove the batch
    // dimension.
```
So variable sequence length does not support deployment with Triton? Or how do I modify the code to make it work?


My client.py
```
import argparse
import numpy as np
import sys
from typing import Dict
import tritonhttpclient
# from tritonclientutils import InferenceServerException

import time
import data_processing
import tokenization
import sys
import codecs
sys.stdout = codecs.getwriter("utf-8")(sys.stdout.detach())

class BertTransformer:
    def __init__(self):
        #  self.model_version = 2
        self.tokenizer = tokenization.FullTokenizer(vocab_file="./vocab.txt", do_lower_case=False)
        
    def preprocess(self, text) -> Dict:
        self.features = data_processing.convert_examples_to_features(text, None, self.tokenizer, max_seq_length=128)
        return self.features
    
bt = BertTransformer()
text = '今天天气不错'
print('text:', text)


features = bt.preprocess(text)



model_name = "varsbert"
triton_client = tritonhttpclient.InferenceServerClient(url='172.17.0.1:9016',verbose=False)
triton_client.is_server_live()



feature = features

S = np.sum(feature["input_mask"])
input_ids = feature["input_ids"][0:S]
segment_ids = feature["segment_ids"][0:S]
cu_seq_lens = np.array([0, S], dtype=np.int32)

max_batch_size = 1
max_sequence_length = 8

shape = (max_sequence_length * max_batch_size, )

inputs = []
outputs = []
inputs.append(tritonhttpclient.InferInput('input_ids', shape, "INT32"))
inputs.append(tritonhttpclient.InferInput('segment_ids', shape, "INT32"))
inputs.append(tritonhttpclient.InferInput('cu_seqlens', (max_batch_size+1,), "INT32"))
inputs.append(tritonhttpclient.InferInput('max_seqlen', (max_sequence_length,), "INT32"))

# Initialize the data
inputs[0].set_data_from_numpy(input_ids.reshape(-1,), binary_data=False)
inputs[1].set_data_from_numpy(segment_ids.reshape(-1,), binary_data=False)
inputs[2].set_data_from_numpy(cu_seq_lens.reshape(-1,), binary_data=False)
inputs[3].set_data_from_numpy(np.array([0 for i in range(max_sequence_length)],dtype=np.int32).reshape(-1,), binary_data=False) # .reshape(-1,)

outputs.append(tritonhttpclient.InferRequestedOutput('bert_out_sum', binary_data=False))
results = triton_client.infer(model_name,inputs,outputs=outputs, model_version='1')
```

**Comment by ttyio** [comment]
@pogevip , hmmm, I have not used triton, not know this limitation before.  
Could you try add batch dims for this input, and modify the plugin code to skip the batch dim? Or maybe people in https://github.com/triton-inference-server/server has a better answer. thanks!

**Comment by pogevip** [comment]
Temporarily closed
